Docs/Features/Logging.md

```diff
@@ -1,27 +1,181 @@
 Logging in P#
 =============
-By default, the P# runtime logger writes all log output to `Console` during production (when verbosity is enabled). During testing, the log output is redirected to an in-memory writer (which dumps it to a file when a bug is found).
+The P# runtime uses the `RuntimeLogWriter` to format and log all runtime messages. To achieve this, it uses the `IRuntimeLogFormatter` interface for formatting messages, and the `ILogger` interface for logging messages.
 
-The runtime logger can be accessed via the following `IMachineRuntime`, `Machine` or `Monitor` property:
-```C#
-ILogger Logger { get; }
-```
+By default, the P# runtime writes all log output to `Console` during production (when verbosity is enabled). During testing, the log output is redirected to an in-memory writer (which dumps it to a file when a bug is found). This behavior can be overriden by changing the default log-related interfaces to custom ones. Further, `RuntimeLogWriter` can be subclassed to change the behavior of how the runtime logs messages.
 
-# Using a custom logger
-It is possible to replace the default P# runtime logger with a custom one that implements the `ILogger` and `IDisposable` interfaces:
+# Using and replacing the logger
+The `ILogger` interface is responsible for writing log messages using the `Write` and `WriteLine` methods. The `IsVerbose` property is `true` when verbosity is enabled, and `false` when disabled.
 ```C#
 public interface ILogger : IDisposable
 {
+  bool IsVerbose { get; set; }
+
   void Write(string value);
+  void Write(string format, object arg0);
+  void Write(string format, object arg0, object arg1);
+  void Write(string format, object arg0, object arg1, object arg2);
   void Write(string format, params object[] args);
```

Comment:
This is because `params object[]` is allocating an object array every single time this method is called, creating unnecessarily allocations and stressing GC! Funnily enough .NET is not automatically optimizing this (not even in Release mode). This was an important optimization that we introduced a few months ago.